File: Recursion/tower_of_hannoi.java
/*
    The Tower of Hanoi problem is a classic problem in computer science and mathematics that involves moving a stack of disks from one peg to another peg. The problem consists of three pegs and a set of disks of different sizes that can slide onto any peg. The objective of the puzzle is to move the entire stack to another peg, obeying the following simple rules:

    Only one disk can be moved at a time.
    Each move consists of taking the upper disk from one of the stacks and placing it on top of another stack or an empty peg.
    No disk may be placed on top of a smaller disk.
    The Tower of Hanoi problem is often used as an example of recursive problem-solving. The solution to the problem can be divided into three parts:

    Move n-1 disks from the starting peg to the auxiliary peg.
    Move the largest disk from the starting peg to the destination peg.
    Move the n-1 disks from the auxiliary peg to the destination peg.
    This process is repeated recursively until all the disks have been moved from the starting peg to the destination peg. The number of moves required to solve the puzzle for n disks can be calculated using the formula 2^n - 1.

    Here's an implementation of the Tower of Hanoi problem in Java using recursion:

*/
public class TowerOfHanoi {
    public static void main(String[] args) {
        int n = 3; // number of disks
        char fromRod = 'A';
        char toRod = 'C';
        char auxRod = 'B';

        System.out.println("Moves to solve Tower of Hanoi problem with " + n + " disks:");
        solveHanoi(n, fromRod, toRod, auxRod);
    }

    public static void solveHanoi(int n, char fromRod, char toRod, char auxRod) {
        if (n == 1) {
            System.out.println("Move disk 1 from rod " + fromRod + " to rod " + toRod);
            return;
        }
        solveHanoi(n - 1, fromRod, auxRod, toRod);
        System.out.println("Move disk " + n + " from rod " + fromRod + " to rod " + toRod);
        solveHanoi(n - 1, auxRod, toRod, fromRod);
    }
}


// -----------------------Explanation------------------------------

// In this implementation, the solveHanoi method is called recursively to move the disks from one rod to another. The method takes in four parameters:

// n: the number of disks
// fromRod: the rod from which the disks are to be moved
// toRod: the rod to which the disks are to be moved
// auxRod: the auxiliary rod which can be used to move the disks
// When the number of disks is 1, the method simply prints the move to be made. Otherwise, the method makes the following three recursive calls:

// Move n-1 disks from fromRod to auxRod using toRod as the auxiliary rod.
// Move the nth disk from fromRod to toRod.
// Move the n-1 disks from auxRod to toRod using fromRod as the auxiliary rod.
// This process is repeated recursively until all the disks have been moved from the starting rod to the destination rod.

File: Recursion/tower_of_hannoi.js
/*
	Towers of Hanoi puzzle.
	Source(https://en.wikipedia.org/wiki/Tower_of_Hanoi)
	Object of the game is to move all the disks over to Tower 3.
	But you cannot place a larger disk onto a smaller disk.

  Approach
	1 Move the top 􀝊 − 1 disks from 􀜵􀝋􀝑􀝎􀜿􀝁 to 􀜣􀝑􀝔􀝅􀝈􀝅􀜽􀝎􀝕 tower,
	2 Move the 􀝊􀯧􀯛 disk from 􀜵􀝋􀝑􀝎􀜿􀝁 to 􀜦􀝁􀝏􀝐􀝅􀝊􀜽􀝐􀝅􀝋􀝊 tower,
	3 Move the 􀝊 − 1disks from 􀜣􀝑􀝔􀝅􀝈􀝅􀜽􀝎􀝕 tower to 􀜦􀝁􀝏􀝐􀝅􀝊􀜽􀝐􀝅􀝋􀝊 tower.
*/
function towerOfHanoi(numDisks, fromPeg, toPeg, auxPeg) {
  if (numDisks === 1) {
    console.log(`Move disk 1 from peg ${fromPeg} to peg ${toPeg}`);
  } else {
    towerOfHanoi(numDisks - 1, fromPeg, auxPeg, toPeg);
    console.log(`Move disk ${numDisks} from peg ${fromPeg} to peg ${toPeg}`);
    towerOfHanoi(numDisks - 1, auxPeg, toPeg, fromPeg);
  }
}

// Example usage:
towerOfHanoi(3, "A", "C", "B"); // Move disk 1 from peg A to peg C, Move disk 2 from peg A to peg B, Move disk 1 from peg C to peg B, Move disk 3 from peg A to peg C, Move disk 1 from peg B to peg A, Move disk 2 from peg B to peg C, Move disk 1 from peg A to peg C

// This implementation uses recursion to solve the problem. The towerOfHanoi function takes four arguments: numDisks (the number of disks to move), fromPeg (the peg the disks start on), toPeg (the peg the disks should end up on), and auxPeg (the auxiliary peg used for moving the disks).
// If numDisks is 1, the function simply moves the single disk from fromPeg to toPeg. Otherwise, it first moves numDisks - 1 disks from fromPeg to auxPeg using toPeg as the auxiliary peg. Then it moves the largest remaining disk from fromPeg to toPeg. Finally, it moves the numDisks - 1 disks from auxPeg to toPeg using fromPeg as the auxiliary peg.
// The function prints out the moves it makes to the console, so running towerOfHanoi(3, "A", "C", "B") would output:

// Move disk 1 from peg A to peg C
// Move disk 2 from peg A to peg B
// Move disk 1 from peg C to peg B
// Move disk 3 from peg A to peg C
// Move disk 1 from peg B to peg A
// Move disk 2 from peg B to peg C
// Move disk 1 from peg A to peg C
